contracts/fair-burn/src/contract.rs

```diff
@@ -0,0 +1,454 @@
+use crate::error::ContractError;
+use crate::msg::{ConfigResponse, ExecuteMsg, InstantiateMsg, QueryMsg, SudoMsg};
+use crate::state::{Config, CONFIG};
+
+use cosmwasm_std::{
+    coin, ensure, to_binary, Addr, BankMsg, Binary, Coin, Decimal, Deps, DepsMut, Env, Event,
+    MessageInfo, StdResult, Uint128,
+};
+use cw2::set_contract_version;
+use cw_utils::maybe_addr;
+use sg_std::{create_fund_fairburn_pool_msg, Response, NATIVE_DENOM};
+use std::collections::BTreeMap;
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:sg-fair-burn";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    let config = Config {
+        fee_percent: Decimal::percent(msg.fee_bps),
+    };
+    config.save(deps.storage)?;
+
+    let event = Event::new("instantiate")
+        .add_attribute("action", "instantiate")
+        .add_attribute("contract_name", CONTRACT_NAME)
+        .add_attribute("contract_version", CONTRACT_VERSION)
+        .add_attribute("fee_percent", config.fee_percent.to_string());
+
+    Ok(Response::new().add_event(event))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    let api = deps.api;
+
+    match msg {
+        ExecuteMsg::FairBurn { recipient } => {
+            execute_fair_burn(deps, info, maybe_addr(api, recipient)?)
+        }
+    }
+}
+
+fn calculate_payouts(funds: &Coin, config: &Config) -> (Coin, Option<Coin>) {
```

Comment:
Been using the convention of putting public functions first so the interface of a contract is more clear. Thoughts?